libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -628,13 +655,72 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	expectDeposit(1, height);
 }
 
+void CodeTransform::setupReturnVariablesAndFunctionExit()
+{
+	yulAssert(!returnVariablesAndFunctionExitAreSetup(), "");
+	yulAssert(m_scope, "");
+
+	ScopeGuard scopeGuard([oldScope = m_scope, this] { m_scope = oldScope; });
+	if (!m_scope->functionScope)
+	{
+		yulAssert(m_scope->superScope && m_scope->superScope->functionScope, "");
+		m_scope = m_scope->superScope;
+	}
+
+	if (m_delayedReturnVariables.empty())
+	{
+		m_functionExitStackHeight = m_evm15 ? 0 : 1;
+		return;
+	}
+
+	// Allocate slots for return variables as if they were declared as variables in the virtual function scope.
+	for (TypedName const& var: m_delayedReturnVariables)
+		(*this)(VariableDeclaration{var.location, {var}, {}});
+
+	m_functionExitStackHeight = ranges::max(m_delayedReturnVariables | ranges::views::transform([&](TypedName const& _name) {
+		return variableStackHeight(_name.name);
+	})) + 1;
+	m_delayedReturnVariables.clear();
+}
+
+namespace
+{
+
+bool statementNeedsReturnVariableSetup(Statement const& _statement, vector<TypedName> const& _returnVariables)
+{
+	if (holds_alternative<FunctionDefinition>(_statement))
+		return true;
```

Comment:
@chriseth I changed this from ``false`` to ``true`` in after the feedback from fuzzing.
This is only hit if there are nested functions, so in practice this shouldn't happen, since we're expected to have optimized up front. But yeah - looking at the ``Block`` visit that calls this, I'm surprised that neither of us realized this - the jump over a nested function will jump over a non-zero stack change if the return arguments haven't been allocated before.

Reply:
Ah nice! I really love the fuzzer!